hugegraph-core/src/main/java/com/baidu/hugegraph/backend/query/ConditionQuery.java

```diff
@@ -522,4 +569,144 @@ private static Object convertNumberIfNeeded(Object value) {
         INDEX,
         INDEX_FILTER
     }
+
+    public static final class Element2IndexValueMap {
+
+        private final Map<Id, Set<LeftIndex>> leftIndexMap;
+        private final Map<Id, Map<Id, Set<Object>>> filed2IndexValues;
+        private Id selectedIndexField;
+
+        public Element2IndexValueMap() {
+            this.filed2IndexValues = new HashMap<>();
+            this.leftIndexMap = new HashMap<>();
+        }
+
+        public void addIndexValue(Id indexField, Id elementId,
+                                  Object indexValue) {
+            if (!this.filed2IndexValues.containsKey(indexField)) {
+                this.filed2IndexValues.put(indexField, new HashMap<>());
+            }
+            Map<Id, Set<Object>> element2IndexValueMap =
+                                 this.filed2IndexValues.get(indexField);
+            if (element2IndexValueMap.containsKey(elementId)) {
+                element2IndexValueMap.get(elementId).add(indexValue);
+            } else {
+                element2IndexValueMap.put(elementId,
+                                          Sets.newHashSet(indexValue));
+            }
+        }
+
+        public void selectedIndexField(Id indexField) {
+            this.selectedIndexField = indexField;
+        }
+
+        public Set<Object> getIndexValues(Id indexField, Id elementId) {
+            if (!this.filed2IndexValues.containsKey(indexField)) {
+                return null;
+            }
+            return this.filed2IndexValues.get(indexField).get(elementId);
+        }
+
+        public void addLeftIndex(Id indexField, Set<Object> indexValues,
+                                 Id elementId) {
+            LeftIndex leftIndex = new LeftIndex(indexValues, indexField);
+            if (this.leftIndexMap.containsKey(elementId)) {
+                this.leftIndexMap.get(elementId).add(leftIndex);
+            } else {
+                this.leftIndexMap.put(elementId, Sets.newHashSet(leftIndex));
+            }
+        }
+
+        public Set<LeftIndex> getLeftIndex(Id elementId) {
+            return this.leftIndexMap.get(elementId);
+        }
+
+        public boolean validRangeIndex(HugeElement element, Condition cond) {
+            // Not UserpropRelation
+            if (!(cond instanceof Condition.UserpropRelation)) {
+                return true;
+            }
+
+            Condition.UserpropRelation propRelation =
+                                       (Condition.UserpropRelation) cond;
+            Id propId = propRelation.key();
+            Set<Object> fieldValues = this.getIndexValues(propId, element.id());
+            if (fieldValues == null) {
+                // Not range index
+                return true;
+            }
+
+            HugeProperty<Object> hugeProperty = element.getProperty(propId);
+            if (hugeProperty == null) {
+                // Property value has been deleted
+                this.addLeftIndex(propId, fieldValues, element.id());
```

Comment:
remove from filed2IndexValues when add LeftIndex to ensure filed2IndexValuesw won't be too large